must-call-qual/src/main/java/org/checkerframework/checker/mustcall/qual/MustCall.java

```diff
@@ -0,0 +1,38 @@
+package org.checkerframework.checker.mustcall.qual;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import org.checkerframework.framework.qual.DefaultFor;
+import org.checkerframework.framework.qual.DefaultQualifierInHierarchy;
+import org.checkerframework.framework.qual.SubtypeOf;
+import org.checkerframework.framework.qual.TypeUseLocation;
+
+/**
+ * If an expression has type {@code @MustCall({"m1", "m2"})}, then its value has an obligation to
+ * eventually call methods {@code m1} and {@code m2} before any exit point or de-allocation.
+ *
+ * <p>The subtyping relationship is:
+ *
+ * <pre>{@code @MustCall({"m1"}) <: @MustCall({"m1", "m2"})}</pre>
+ *
+ * <p>For example, an object with an obligation to call only "m1" can be passed to a method that
```

Comment:
This type system is unorthodox, because it is not tracking a fact about the values in the program directly. Rather, it is tracking an obligation for each value that is a superset of what it should do before it is de-allocated.

An object with the type `@MustCall({"m1", "m2"}) Object`, just as `param` in your example, has an obligation to call `m1()` and `m2` before it is de-allocated. This says nothing about what the object might actually contain; it is stating a fact that is otherwise external to the existing type system. In other words, this type system in a sense tracks provenance (like the tainting checker) rather than the possible values that an object might take on (like the value checker).

Maybe the confusion here stems from the fact that this checker doesn't actually enforce this property directly; like the Value Checker, it is intended only to provide information to other checkers (in this case, the must-call analysis embedded in the Object Construction Checker proper). It's worth noting that that analysis *isn't* a type system; rather, it's a dataflow analysis that runs directly on the CFG after both the MustCall checker here and the Called Methods checker. The facts that those two checkers produce allow that dataflow analysis to do its job, which is to ensure that each MustCall obligation has been fulfilled.

Another way to think about this is that `@MustCall({"m1", "m2"}) Object` represents all objects that need to call `m1`, `m2`, both, or neither. It cannot represent an object that needs to call some *other* method. So, for example, `@MustCall({"m1", "m2"}) Object` represents more possible values than `@MustCall({"m1"}) Object` does, because `@MustCall({"m1"}) Object` can only contain an object that needs to call `m1` or one that needs to call nothing - an object that needs to call `m2` (or both `m1` and `m2`) is not permitted. This view makes the need for `@MustCallTop` more apparent: `@MustCallTop` represents all objects, which might need to call any set of their methods.

Hopefully this makes the design of the type system make sense. I'm sure we can find ways to make the documentation both clearly express this and more concise than what I've written above. I've made an attempt to rewrite the definition of the `@MustCall` annotation. Can you take a look at let me know if that makes it more clear?

> Maybe you can recast the annotation as `@WillCall`.

I don't think that's quite right, either, but I don't have a better suggestion. `@MustCall` makes intuitive sense to me (and I think it did to you, too, when we discussed the idea on the phone a few weeks ago), but I agree that it doesn't quite capture the notion of what I've described in the preceding paragraphs.

Reply:
> Another way to think about this is that `@MustCall({"m1", "m2"}) Object` represents all objects that need to call `m1`, `m2`, both, or neither. It cannot represent an object that needs to call some _other_ method.

This is a quite different definition than was previously documented.  Thanks for the update; that helps.

Maybe a name like `MightMustCall` or `MustCallAtMost` would better convey this new semantics.


Your arguments about being unorthodox or not enforcing or not being a type system don't make sense to me.  Even if this were only a dataflow analysis, it would still need a clear definition of the dataflow lattice, in which lower lattice elements represent fewer values.  Hopefully we are converging on that, so that is good news.

Reply:
> This is a quite different definition than was previously documented. Thanks for the update; that helps.

Thanks for helping me make the documentation match what the code actually does!

> Maybe a name like MightMustCall or MustCallAtMost would better convey this new semantics.

I think either of those names would be more verbose than is necessary, and would be confusing to users. The colloquial definition of what `obj must call foo` means is that `obj` needs to call `foo` along all paths, even if on some paths `obj` contains some dynamic object that doesn't really need to call `foo` (or at least that's what I mean when I say such things, and I think @msridhar and @Nargeshdb agree - though they should point out if they do not). The semantics of the `@MustCall` annotation match this meaning.

> Your arguments about being unorthodox or not enforcing or not being a type system don't make sense to me. Even if this were only a dataflow analysis, it would still need a clear definition of the dataflow lattice, in which lower lattice elements represent fewer values. Hopefully we are converging on that, so that is good news.

Fair enough. Those arguments were intended to help explain why this was difficult to explain, but they might have just made this more confusing in retrospect.